src/ht.c

```diff
@@ -1,239 +1,238 @@
+// size - number of buckets in the hashtable
+// hashfunction - the function that does the hashing, must not be null.
+// comparator - the function to check if values are equal, if NULL, just checks == (for storing ints).
+// keydup - function to duplicate to key (eg strdup), if NULL just does =.
+// valdup - same as keydup, but for values
+// pair_free - function for freeing a keyvaluepair - if NULL just does free.
+// calcsize - functino to calculate the size of a value. if NULL, just stores 0.
+SdbHash* internal_ht_new(ut32 size, HashFunction hashfunction, ListComparator comparator, DupKey keydup, DupValue valdup, HtKvFreeFunc pair_free, CalcSize calcsize) {
+	SdbHash* ht;
+	ut32 i;
+	ht = calloc (1, sizeof (*ht));
+	if (ht) {
+		ht->size = size;
+		ht->count = 0;
+		ht->prime_idx = 0;
+		ht->load_factor = 1;
+		ht->hashfn = hashfunction;
+		ht->cmp = comparator;
+		ht->dupkey = keydup;
+		ht->dupvalue = valdup;
+		ht->table = calloc (ht->size, sizeof (SdbList*));
+		ht->calcsize = calcsize;
+		if (ht->table) {
+			for (i = 0; i < ht->size; ++i) {
+				ht->table[i] = ls_newf ((SdbListFree)pair_free);
+			}
+		} else {
+			free (ht);
+			return NULL;
+		}
+	}
+	return ht;
+}
 
-#define ARRAY_SIZE(array) (sizeof(array) / sizeof(array[0]))
-
-/*
- * From Knuth -- a good choice for hash/rehash values is p, p-2 where
- * p and p-2 are both prime.  These tables are sized to have an extra 10%
- * free to avoid exponential performance degradation as the hash table fills
- */
-
-static ut32 deleted_data;
-
-static const struct {
-	ut32 max_entries, size, rehash;
-} hash_sizes[] = {
-	{ 2,		5,		3	  },
-	{ 4,		7,		5	  },
-	{ 8,		13,		11	  },
-	{ 16,		19,		17	  },
-	{ 32,		43,		41        },
-	{ 64,		73,		71        },
-	{ 128,		151,		149       },
-	{ 256,		283,		281       },
-	{ 512,		571,		569       },
-	{ 1024,		1153,		1151      },
-	{ 2048,		2269,		2267      },
-	{ 4096,		4519,		4517      },
-	{ 8192,		9013,		9011      },
-	{ 16384,	18043,		18041     },
-	{ 32768,	36109,		36107     },
-	{ 65536,	72091,		72089     },
-	{ 131072,	144409,		144407    },
-	{ 262144,	288361,		288359    },
-	{ 524288,	576883,		576881    },
-	{ 1048576,	1153459,	1153457   },
-	{ 2097152,	2307163,	2307161   },
-	{ 4194304,	4613893,	4613891   },
-	{ 8388608,	9227641,	9227639   },
-	{ 16777216,	18455029,	18455027  },
-	{ 33554432,	36911011,	36911009  },
-	{ 67108864,	73819861,	73819859  },
-	{ 134217728,	147639589,	147639587 },
-	{ 268435456,	295279081,	295279079 },
-	{ 536870912,	590559793,	590559791 },
-	{ 1073741824,	1181116273,	1181116271},
-	{ 2147483648ul,	2362232233ul,	2362232231ul}
-};
+bool ht_delete_internal(SdbHash* ht, const char* key, ut32* hash) {
+	ut32 bucket;
+	SdbListIter* iter = NULL;
+	SdbList* list = NULL;
+	SdbKv* kvp = NULL;
+	ut32 computed_hash;
+	if (!hash) {
+		computed_hash = ht->hashfn (key);
+	} else {
+		computed_hash = *hash;
+	}
+	bucket = computed_hash % ht->size;
+	list = ht->table[bucket];
+	ls_foreach (list, iter, kvp) {
```

Comment:
iirc ls_foreach already checks for null, sot he previous initializer statement will not be necesary